Tests/ApolloTests/PossiblyDeferredTests.swift

```diff
@@ -0,0 +1,390 @@
+  
+  func testErrorReturnedFromFlatMapIsPropagated() {
+    let possiblyDeferred = PossiblyDeferred.deferred { "foo" }
+      .flatMap { _ -> PossiblyDeferred<String> in .immediate(.failure(TestError())) }
+    
+    XCTAssertThrowsError(try possiblyDeferred.get()) { error in
+      XCTAssert(error is TestError)
+    }
+  }
+
+  // MARK: - Map error
+  
+  func testMapErrorOverImmediateFailure() {
+    let possiblyDeferred = PossiblyDeferred<String>.immediate(.failure(TestError()))
+      .mapError { error in
+        XCTAssert(error is TestError)
+        return OtherTestError()
+      }
+    
+    XCTAssertThrowsError(try possiblyDeferred.get()) { error in
+      XCTAssert(error is OtherTestError)
+    }
+  }
+  
+  func testMapErrorOverDeferredFailure() {
+    let possiblyDeferred = PossiblyDeferred<String>.deferred { throw TestError() }
+      .mapError { error in
+        XCTAssert(error is TestError)
+        return OtherTestError()
+      }
+    
+    XCTAssertThrowsError(try possiblyDeferred.get()) { error in
+      XCTAssert(error is OtherTestError)
+    }
+  }
+  
+  func testMapErrorOverMapOverImmediateFailure() {
+    let possiblyDeferred = PossiblyDeferred<String>.immediate(.failure(TestError()))
+      .map { _ in "foo" }
+      .mapError { error in
+        XCTAssert(error is TestError)
+        return OtherTestError()
+      }
+    
+    XCTAssertThrowsError(try possiblyDeferred.get()) { error in
+      XCTAssert(error is OtherTestError)
+    }
+  }
+  
+  func testMapErrorOverMapOverDeferredFailure() {
+    let possiblyDeferred = PossiblyDeferred<String>.deferred { throw TestError() }
+      .map { _ in "foo" }
+      .mapError { error in
+        XCTAssert(error is TestError)
+        return OtherTestError()
+      }
+    
+    XCTAssertThrowsError(try possiblyDeferred.get()) { error in
+      XCTAssert(error is OtherTestError)
+    }
+  }
+  
+  func testMapErrorOverMapThrowingErrorOverImmediateSuccess() {
+    let possiblyDeferred = PossiblyDeferred.immediate(.success("foo"))
+      .map { value -> String in
+        throw TestError()
+      }
+      .mapError { error in
+        XCTAssert(error is TestError)
+        return OtherTestError()
+      }
+    
+    XCTAssertThrowsError(try possiblyDeferred.get()) { error in
+      XCTAssert(error is OtherTestError)
+    }
+  }
+  
+  func testMapErrorOverMapThrowingErrorOverDeferredSuccess() {
+    let possiblyDeferred = PossiblyDeferred.deferred { "foo" }
+      .map { value -> String in
+        throw TestError()
+      }
+      .mapError { error in
+        XCTAssert(error is TestError)
+        return OtherTestError()
+      }
+    
+    XCTAssertThrowsError(try possiblyDeferred.get()) { error in
+      XCTAssert(error is OtherTestError)
+    }
+  }
+  
+  // MARK: - Lazily evaluate all
+  
+  func testLazilyEvaluateAllIsDeferred() throws {
+    let possiblyDeferreds: [PossiblyDeferred<String>] = [.deferred { "foo" }, .deferred { "bar" }]
+
+    var numberOfInvocations = 0
```

Comment:
all these `numberOfInvocations` tests make me wonder: Are we doing anything to limit things to one invocation? or is this something where the number of invocations could be arbitrary?

Reply:
The number of invocations could be arbitrary, so any caching you want to do would happen downstream (in `DataLoader` for example). I thought about making `PossiblyDeferred` cache the result by changing itself from `deferred` to `immediate` the first time the deferred result gets resolved. But that won't really work because it is a value type. And changing it to a class makes these more heavy weight than I wanted.

Reply:
True - I think i'm more thinking what happens if you accidentally call `get` twice - you'd get `foobarbar`, right? I guess at that point the format of the incorrect data *should* be a giveaway, but if this is in a big chain it could be difficult to tell where you accidentally called something twice.

Reply:
Ah, you're talking specifically about what happens with the test. We also have assertions for `numberOfInvocations`, so that should tell you what happened?

Reply:
No - I was using the test setup as an example. I'm wondering what happens if you call it twice in general. It's clear that _in the test_ it's only getting called once.

Reply:
I'm not sure I know what you mean. A `PossiblyDeferred` that's `.deferred` can be invoked multiple times, that's by design and should not be an issue. It's similar to how you could call `result.get()` multiple times.